### DISCUSSION TITLE
[Feature]: Vercel Hybrid Edge and Serverless Deployment

### DISCUSSION THREAD

**Discussion by carlassmann** [remix-run/remix#2884] [Proposals]
### What is the new or updated feature that you are suggesting?

Currently, remix deploys the entire app to Vercel Serverless Functions using Vercels [File System API](https://vercel.com/docs/file-system-api). 

We could allow deploying the entire app to Vercel Edge Functions as well, which is an abstraction over Cloudflare Workers.

An even better way to take advantage of Vercel would be an optional hybrid approach, where we can choose per route (or similar) if this should run in a Serverless Function or in an Edge Function. 

### Why should this feature be included?

Remix made clear that SSR on the Edge is the future. Vercel allows deploying to the Edge, and we should support it.

Some functionality might still rely on NodeJS (e.g. `nodemailer`, `firebase-admin`). For these situations, a hybrid approach is the killer feature, since there are likely a few routes, which can still run on the edge.

**Comment by swamidass** [comment]
I agree. This would be good.

At the moment, which of the deploy options do go to the edge? Clarity in the documentation, or a way of querying a project config/etc would be good to know.

**Comment by carlassmann** [comment]
Generally speaking:

Remix compiles the whole app to one single request handler. Developers can use this handler via adapters to deploy their app anywhere (through an adapter). What Remix would need is an option to return multiple request handlers based on some flag. Then, developers can use different request handlers at different places of their app.

After that, the Vercel adapter could deploy one handler to an edge function and another one to serverless functions.

**Comment by cliffordfajardo** [comment]
> Remix compiles the whole app to one single request handler. 

@ccssmnn - is this always the case even when you're deploying a non-serverless environment? Ex: AWS, Azure, Digital Ocean etc?

Would the scope of this change only require changes to the vercel adapter or is this general high level change to remix?
- https://remix.run/docs/en/v1/other-api/adapter#server-adapters

**Comment by sergiodxa** [comment]
> I agree. This would be good.
> 
> At the moment, which of the deploy options do go to the edge? Clarity in the documentation, or a way of querying a project config/etc would be good to know.

Cloudflare deploys to the edge.

If you deploy to Fly.io you could also run your app on different servers around the world, not as many as Cloudflare

**Comment by carlassmann** [comment]
@cliffordfajardo the way Remix builds the app is independent of the adapter used. See [the Technical Explanation](https://remix.run/docs/en/v1/pages/technical-explanation#http-handler-and-adapters).

**Comment by swamidass** [comment]
Perhaps, but look at the file API from Vercel here: https://vercel.com/docs/file-system-api

Currently Remix puts the server file into: "api/server.js". Which is NOT run at the edge. If the file is placed at "api/_middleware.js", however, it might run at the edge. 

Perhaps...

1.  Remix needs APIs or convention available for deciding which code our routes should run at the edge or not in Remix. Perhaps a "middleware" or "edge" hook in the route files? 
 
2. Specifically for Vercel, a variant or config option that deploys to api/_middleware.js, not api/server.js, much like there is two cloudflare versions of the deploy.

It seems #1 is important because there are tradeoffs between centralized vs. on the edge, and many applications may need to have control over what runs where.

**Comment by egoist** [comment]
For your information SvelteKit now supports deploying the entire app as a Vercel Edge Function (as an option), which could be an option for Remix, but I also really want the ability to continue deploying my Remix app as a serverless function (so it's closer to my database) and use a middleware (akin to _middleware.ts in Next.js) instead to handle some tasks like bot detection and auth redirect

**Comment by impactvelocity** [comment]
How would this work with Remix:

https://vercel.com/docs/concepts/functions/edge-functions 

`
export const config = {
  runtime: 'experimental-edge',
};

`

**Comment by JoeMethven** [comment]
Any update on this? Is Vercel edge functions supported with remix?

**Reply by markmals** [reply]
There’s still no official support or adapter, but [it doesn’t look too difficult to hack together your own support using the Cloudflare adapter](https://github.com/mcansh/remix-vercel-edge)

**Comment by itsMapleLeaf** [comment]
I haven't tried [this hack](https://github.com/remix-run/remix/discussions/2884#discussioncomment-3626685) but I might dig into that later. Official support would still be really nice to see, though!

**Comment by alfl** [comment]
As SSR imports get bigger and take longer, CF Workers fail to start. This will be a limiting factor that Remix can (partly) help mitigate.

Related: https://github.com/remix-run/remix/discussions/5226

**Comment by WesleyYue** [comment]
They're working on it! https://github.com/vercel/vercel/pull/9375

**Comment by brophdawg11** [comment]
This discussion is being closed as part of the Remix v3 repository cleanup.

Remix v1/v2 was [upstreamed into React Router v7](https://remix.run/blog/wake-up-remix), so if this proposal is still relevant, please reopen or continue the discussion in the [React Router](https://github.com/remix-run/react-router) repository.
